Release Docker image for linux/arm64, linux/amd64 on Travis CI

odidev opened the pull request:
The following file has been created and modified:
Added support of buildx in .travis.yml to build and push the docker image for arm64 and amd64 platforms and updated the distribution to focal.
Updated the dockerfile to build and push the images for latest versions to dockerhub using the ARG TARGETARCH.

Signed-off-by: odidev <odidev@puresoftware.com>

PeterDaveHello commented on .travis.yml:
```diff
@@ -1,6 +1,6 @@
 ---
 
-dist: bionic
+dist: focal
```
Is this a required change?

odidev replied:
Yes, when we update the distribution to focal there is no need to install the latest version of docker separately as it is required to use buildx commands.

PeterDaveHello replied:
Okay, I got it, if you don't mind, another separated PR can also be sent to upgrade the Ubuntu distro on Travis CI, as it's not a blocking issue to the Docker build, and the distro will be upgraded anyway.

PeterDaveHello commented on .travis.yml:
```diff
@@ -36,7 +36,15 @@ jobs:
-        - docker build --build-arg AZCOPY_VERSION="$LATEST_AZCOPY" -t docker-azcopy:$TRAVIS_COMMIT .
+
+        - docker login -u ${DOCKER_USERNAME} -p ${DOCKER_PASSWORD}
+        - travis_wait 30 docker buildx build --build-arg AZCOPY_VERSION="$LATEST_AZCOPY" -t $DOCKER_USERNAME/docker-azcopy:$TRAVIS_COMMIT --platform linux/amd64,linux/arm64 --push . >> /dev/null 2>&1
```
What's the travis_wait here for?

odidev replied:
Firstly, I was facing max size exceeded error with huge amount of logs. So, I have redirected logs to /dev/null and after this build was running for more than 10 mins and as per the travis, build gets terminated if build is stalled at a command for more than 10 mins. Hence, I have increased waiting time to 30 mins using travis_wait.

PeterDaveHello replied:
I wonder in this way if we can't know what's going on when something went wrong?

odidev commented:
@PeterDaveHello,

Rebased and merged the commit. Please let me know if any more changes are required.

PeterDaveHello commented:
@odidev thanks, I'll take a look soon!

PeterDaveHello reviewed (changes requested):
@odidev I believe it's almost done, just found that some logic is still missing, the original logic is:

https://github.com/PeterDaveHello/docker-azcopy/blob/master/hooks/build#L3-L7

Which means when it's tagged, the `$LATEST_AZCOPY` should be the same as the tag, when it's not tagged but it's the build of master branch, the `$LATEST_AZCOPY` should be the latest azcopy version, fetched from GitHub API

odidev commented:
@PeterDaveHello,
I have implemented the requested changes and sharing you a commit for the same. 
I have added “DOCKER_TAG” and “SOURCE_BRANCH” in travis config/secrets and tested it. Please have a look and share your thoughts.

Commit link: https://github.com/odidev/docker-azcopy/commit/ea7c8b52bf04034c5630157649798617a19b1c24

If this is exactly what you have suggested, I will amend the commit. If not, please suggest on what I must add more.

PeterDaveHello commented:
@odidev Sure, thanks!

PeterDaveHello commented:
@odidev please take a look ;)
https://github.com/odidev/docker-azcopy/commit/ea7c8b52bf04034c5630157649798617a19b1c24#r60156035

odidev commented:
@PeterDaveHello, I have implemented the requested changes and as per your suggestion I have removed “DOCKER_TAG” and now the if condition is getting validated only through the “SOURCE_BRANCH” in travis config/secrets and I have tested the same. Please review and share your thoughts on the same.

PeterDaveHello commented:
Oops, some original blank lines in `.travis,yml` are now also removed, which caused the conflicts in `.travis.yml` that make this PR not able to be merged 😱

odidev commented:
@PeterDaveHello 
The conflicts has been removed and tested the same which works fine from my end. Please review the same.

PeterDaveHello commented:
@odidev looks good to me, but the CI failed, I'll take a look.

PeterDaveHello commented on .travis.yml:
```diff
@@ -33,10 +33,24 @@ jobs:
+            AZCOPY_VERSION=$(curl -sLo- https://api.github.com/repos/Azure/azure-storage-azcopy/releases | grep tag_name | head -n 1 | awk -F'"' '{print $4}' | sed 's/^v//g');
+          else
+            AZCOPY_VERSION="$SOURCE_BRANCH";
+          fi
```
Looks like we have a problem here.

According to the docs: https://docs.travis-ci.com/user/environment-variables/, we should use `$TRAVIS_BRANCH`, right?

> - for push builds, or builds not triggered by a pull request, this is the name of the branch.
> - for builds triggered by a pull request this is the name of the branch targeted by the pull request.
> - for builds triggered by a tag, this is the same as the name of the tag (`TRAVIS_TAG`).

Using `AZCOPY_VERSION="$TRAVIS_BRANCH"` will still have problem for pull requests.

What about to check `$TRAVIS_TAG` first, assign it to `$AZCOPY_VERSION` when it's existed? When it doesn't exist, we use `curl` to determinate the version.

PeterDaveHello replied:
@odidev this one is not changed since the last few review.

PeterDaveHello commented on .travis.yml:
```diff
@@ -33,10 +33,24 @@ jobs:
+          else
+            AZCOPY_VERSION="$SOURCE_BRANCH";
+          fi
+        - export AZCOPY_VERSION
```
Doesn't look like we need to export it, but I believe that we can print (`echo`) it to help we debug.

PeterDaveHello reviewed (changes requested):
Hi @odidev I think I found the problem, still the similar issue, `$SOURCE_BRANCH` is on Docker Hub, not Travis CI, please take a look at my comment ;)

PeterDaveHello commented:
hmmm ... conflicted again 🤔

odidev commented:
@PeterDaveHello 
Resolved the conflict, please check and let me know your review for the same.

PeterDaveHello commented:
@odidev the above mentioned issue doesn't seem to be changed, not sure if the wrong commit was pushed?

odidev commented:
@PeterDaveHello Understood the requirements and implemented in such a way that if we push from master branch the AZCOPY_VERSION would be determined by using curl, else the TRAVIS_TAG would be assigned as AZCOPY_VERSION.
Please have a look and share your thoughts on the same.

PeterDaveHello commented:
Hi @odidev,

I saw that you changed a `$AZCOPY_VERSION` assignment from `$TRAVIS_BRANCH` to `$TRAVIS_TAG` 👍 

It's interesting that the build didn't fail, because the build of this pull request is not a tagged build nor a master branch build, what about to check $TRAVIS_TAG first, assign it to $AZCOPY_VERSION when it's existed? When it doesn't exist, we use curl to determinate the version. Besides the master branch and tagged branch, we'll still need to make sure other branches, like those from pull request, to build the latest stable version of azcopy.

BTW, maybe this one could also be updated next time: https://github.com/PeterDaveHello/docker-azcopy/pull/4#discussion_r762776134 ;)

PeterDaveHello commented:
Hi @odidev, sorry that the process is a little bit long, I'm ready to get this merged, and release the arm image from the CI, instead of my manually build! We're so close to it, but if you don't have enough time or have some difficulty, please feel free to let me know. Thanks a lot for the contribution 👍

odidev commented:
Hi @PeterDaveHello 

I'm happy to know that you are ready to merge my contribution. 

As per the discussion I have implemented  in a way where pull request from other branches than master would check for TRAVIS_TAG and made it to assign the same to  AZCOPY_VERSION.
I have echo the AZCOPY_VERSION instead of exporting it. 

So, I feel the requirements are met, else please give me some more clarity on the same.

PeterDaveHello commented:
Hi @odidev,

Looks this place is still not changed: https://github.com/PeterDaveHello/docker-azcopy/pull/4/files#r762775753, 

Though I don't know what's the reason why the build didn't fail, the build of this pull request is not a tagged build nor a master branch build, `$TRAVIS_TAG` should be empty in this case, unless I missed anything, or Travis CI's document is dated.

I think we need to check `$TRAVIS_TAG` first, assign it to `$AZCOPY_VERSION` when `$TRAVIS_TAG` exists, and when it doesn't exist, we use curl to determinate the version.

odidev commented:
@PeterDaveHello 
I hope this is the logic which you required for, I have implemented in a way where 

1. _**TRAVIS_TAG doesn't exists:**_

    if **TRAVIS_TAG**  it doesn't exist or empty it will use curl to determine the **AZCOPY_VERSION.** 
    
2. _**TRAVIS_TAG exists:**_

    if **TRAVIS_TAG**  exists AZCOPY_VERSION would be the proper **TRAVIS_TAG.** 
    
 >>Though I don't know what's the reason why the build didn't fail, the build of this pull request is not a tagged build nor a master branch build, $TRAVIS_TAG should be empty in this case, unless I missed anything, or Travis CI's document is dated.
 
I have checked the PR builds of your repo for the last commit, it says the PR is from master branch in TravisCI   job description for the same. please refer this for more details : https://app.travis-ci.com/github/PeterDaveHello/docker-azcopy/builds/243787441

PeterDaveHello commented:
@odidev sorry for the late review, the master branch should be the targeted branch, I missed it. The latest change looks good enough to be merged already, thanks a lot!

PeterDaveHello reviewed (approved):
Happy New Year 🎉

odidev commented:
> Happy New Year 🎉

Thanks :)
Happy New Year 🎉
